# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/7ayder-99/flyrank_internship/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I chose the Refresh / Content Opportunity Scoring lane because it focuses on a real content decision: identifying which pages should be reviewed first when resources are limited. The starter pipeline already demonstrates a ranking workflow using observable search and engagement signals, making this lane a natural extension of the work completed in the starter notebooks. My goal is to build a decision-support system that helps prioritize content review opportunities based on evidence rather than simple rules.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

Which pages should be reviewed first for refresh, expansion, protection, or monitoring based on observable search and engagement signals?

Decision

The decision is how to prioritize content review efforts across a large set of pages.

Action

A content or SEO team can review the highest-ranked pages first and decide whether to refresh, expand, monitor, or protect them.

Cost of a Wrong Recommendation

False Positive:
A page is recommended for review even though it is not a meaningful opportunity. This wastes content, SEO, and analyst time.

False Negative:
A valuable opportunity is missed. A page that could have benefited from review remains unchanged, potentially resulting in lost traffic or engagement.

Why Data and ML Can Help

Content performance depends on multiple factors such as impressions, clicks, engagement, freshness, position, and traffic trends. Machine learning may help combine these signals into a more useful ranking than a single rule or metric.

In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [4]:
import os, sys, subprocess

IN_COLAB = "google.colab" in sys.modules
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if IN_COLAB:
    if not os.path.isdir(REPO_DIR):
        subprocess.run(["git", "clone", "--depth", "1", REPO_URL, REPO_DIR], check=True)
    os.chdir(REPO_DIR)
    subprocess.run([sys.executable, "-m", "pip", "install", "-q", "-r", "requirements.txt"], check=True)
else:
    # find the repo root from wherever this kernel started
    while not os.path.isdir("data/raw") and os.getcwd() != "/":
        os.chdir("..")

print("Working dir:", os.getcwd())
assert os.path.exists("data/raw/content_refresh_anonymized.csv"), "starter CSV not found — are you at the repo root?"
print("Starter data found. You're ready.")


Working dir: /content/flyrank-ml-internship-starter
Starter data found. You're ready.


In [5]:
import pandas as pd

df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

print("Number of pages:", len(df))
print("Pages with impressions > 0:", (df["impressions_90d"] > 0).sum())
print("Average impressions (90d):", round(df["impressions_90d"].mean(), 2))
print("Average sessions (90d):", round(df["sessions_90d"].mean(), 2))

Number of pages: 30000
Pages with impressions > 0: 30000
Average impressions (90d): 5200.37
Average sessions (90d): 37.07


The starter dataset contains 30000 pages.

30000 pages received at least one search impression during the 90-day window.

The average page received 5200.37 impressions and 37.07 sessions during the same period.

These numbers suggest there is meaningful variation in page performance, making content prioritization a useful decision-support problem for further analysis.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

----What I Can Claim

-I can identify observed patterns in the available data.

-I can estimate which pages appear more likely to deserve review.

-I can build a decision-support ranking based on measured signals.

-I can compare model performance against a transparent baseline.

----What I Cannot Claim

-I cannot prove that refreshing a page will cause traffic growth.

-I cannot prove Google ranking factors.

-I cannot guarantee future performance.

-I cannot claim causal relationships from observational data alone.

I cannot replace human review and judgment.


In [ ]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.